Source/InvocationCollection.cs

```diff
@@ -47,20 +47,20 @@ namespace Moq
 {
 	internal sealed class InvocationCollection : IReadOnlyList<IReadOnlyInvocation>
 	{
-		private List<Invocation> invocations;
+		private Invocation[] invocations;
 
-		public InvocationCollection()
-		{
-			this.invocations = new List<Invocation>();
-		}
+		private int capacity = 0;
+		private int count = 0;
+
+		private readonly object invocationsLock = new object();
 
 		public int Count
 		{
 			get
 			{
-				lock (this.invocations)
+				lock (this.invocationsLock)
 				{
-					return this.invocations.Count;
+					return count;
 				}
```

Comment:
I believe this `lock` is superfluous... please remove. Reading an `int` is an atomic operation, there's no danger of struct tearing or anything of the sort. If there is a concurrent mutation happening during a read, it doesn't really matter whether external code sees the original or updated `count`.

(In fact, we could question the whole point of having a `Count` property in a concurrent collection, since user code isn't supposed to do non-atomic accesses like `if (invocations.Count > 0) firstInvocation = invocations[0];`... but having `Count` can still be useful in single-threaded scenarios.)

Reply:
It's not there to avoid a torn read, but to protect against reordering optimisations and variable caching. If you want to go down that avenue, it might require marking the backing field as volatile, or otherwise setting up a memory barrier. I don't advise pulling on this thread unless you think it's very important.

Reply:
I didn't know reordering happened across different methods. But yeah, this isn't super-important. Let's leave it as is, then.